File: backend/python_microservices/caption_service.py
import os
from flask import Flask, request, jsonify
from groq import Groq
from dotenv import load_dotenv
from pathlib import Path

# Load env from parent directory
env_path = Path(__file__).resolve().parent.parent / '.env'
load_dotenv(dotenv_path=env_path)

app = Flask(__name__)
client = Groq(api_key=os.environ.get("GROQ_API_KEY"))

@app.route('/process', methods=['POST'])
def process():
    print("--- Python Caption Service Triggered ---")
    data = request.json
    image_url = data.get('imageUrl')
    
    if not image_url:
        return jsonify({"error": "imageUrl is required"}), 400

    try:
        completion = client.chat.completions.create(
            model="meta-llama/llama-4-scout-17b-16e-instruct",
            messages=[
                {
                    "role": "user",
                    "content": [
                        {
                            "type": "image_url",
                            "image_url": {"url": image_url}
                        },
                        {
                            "type": "text",
                            "text": "This is a product image. Describe what you see in 1-2 short sentences. Focus on: product type, brand name if visible, color, and key features like SPF or ingredients. Be specific and concise — this description will be used as a search query."
                        }
                    ]
                }
            ],
            max_tokens=200,
        )
        result = completion.choices[0].message.content.strip()
        print(f"Result: {result}")
        return jsonify({"result": result})
    except Exception as e:
        print(f"Error: {e}")
        return jsonify({"error": str(e)}), 500

if __name__ == '__main__':
    app.run(port=5001)

File: frontend/microservices/caption-service.js
import express from 'express';
import dotenv from 'dotenv';
import path from 'path';
import { fileURLToPath } from 'url';
import { getCaption } from '../lib/services/groqService.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Load env from project root
dotenv.config({ path: path.join(__dirname, '../.env.local') });

const app = express();
app.use(express.json());

app.post('/process', async (req, res) => {
  console.log('--- Caption Service Triggered ---');
  const { imageUrl } = req.body;
  
  if (!imageUrl) return res.status(400).json({ error: 'imageUrl is required' });

  try {
    const result = await getCaption(imageUrl);
    console.log('Result:', result);
    res.json({ result });
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: error.message });
  }
});

const PORT = 3001;
app.listen(PORT, () => {
  console.log(`\x1b[36m[Caption Service]\x1b[0m running on http://localhost:${PORT}`);
});
